File: apps/api/internal/handlers/teams/invitations.go
package teams

import (
	"time"

	validation "github.com/go-ozzo/ozzo-validation/v4"
	"github.com/go-ozzo/ozzo-validation/v4/is"
	"github.com/gofiber/fiber/v3"
	"github.com/ted-too/logsicle/internal/storage"
	"github.com/ted-too/logsicle/internal/storage/models"
)

// CreateInvitationRequest represents the request body for creating an invitation
type CreateInvitationRequest struct {
	Email string      `json:"email"`
	Role  models.Role `json:"role"`
}

func (r CreateInvitationRequest) Validate() error {
	return validation.ValidateStruct(&r,
		validation.Field(&r.Email, validation.Required, is.Email),
		validation.Field(&r.Role, validation.Required, validation.In(
			models.RoleAdmin,
			models.RoleMember,
		)),
	)
}

// CreateInvitation creates a new invitation for a user to join an organization
func (h *TeamsHandler) CreateInvitation(c fiber.Ctx) error {
	userSession := c.Locals("session").(storage.Session)

	// Check if active organization is set
	if userSession.ActiveOrganization == "" {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error":   "No active organization",
			"message": "You must have an active organization to create an invitation",
		})
	}

	var organization models.Organization
	if err := h.db.First(&organization, "id = ?", userSession.ActiveOrganization).Error; err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error":   "Failed to load organization",
			"message": err.Error(),
		})
	}

	var input CreateInvitationRequest
	if err := c.Bind().Body(&input); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error":   "Invalid request body",
			"message": err.Error(),
		})
	}

	if err := input.Validate(); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error":   "Validation failed",
			"message": err.Error(),
		})
	}

	// Check if the user is already a member of the organization
	var existingMembership models.TeamMembership
	err := h.db.Where("organization_id = ? AND user_id IN (SELECT id FROM users WHERE email = ?)",
		userSession.ActiveOrganization, input.Email).First(&existingMembership).Error
	if err == nil {
		return c.Status(fiber.StatusConflict).JSON(fiber.Map{
			"error":   "User is already a member",
			"message": "This user is already a member of the organization",
		})
	}

	// Check if there's a pending invitation for this email
	var existingInvitation models.Invitation
	err = h.db.Where("organization_id = ? AND email = ? AND status = ?",
		userSession.ActiveOrganization, input.Email, models.InvitationPending).First(&existingInvitation).Error
	if err == nil {
		return c.Status(fiber.StatusConflict).JSON(fiber.Map{
			"error":   "Invitation already exists",
			"message": "There is already a pending invitation for this email",
		})
	}

	// Create the invitation
	invitation := models.Invitation{
		Email:          input.Email,
		OrganizationID: userSession.ActiveOrganization,
		Role:           input.Role,
		Status:         models.InvitationPending,
		ExpiresAt:      time.Now().Add(7 * 24 * time.Hour), // 7 days
		InvitedByID:    userSession.UserID,
	}

	if err := h.db.Create(&invitation).Error; err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error":   "Failed to create invitation",
			"message": err.Error(),
		})
	}

	// TODO: Send invitation email with the token link

	return c.Status(fiber.StatusCreated).JSON(invitation)
}

// ListInvitations lists all invitations for the current organization
func (h *TeamsHandler) ListInvitations(c fiber.Ctx) error {
	userSession := c.Locals("session").(storage.Session)

	// Check if active organization is set
	if userSession.ActiveOrganization == "" {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error":   "No active organization",
			"message": "You must have an active organization to list invitations",
		})
	}

	var invitations []models.Invitation
	if err := h.db.
		Preload("InvitedBy").
		Preload("Organization").
		Where("organization_id = ?", userSession.ActiveOrganization).
		Find(&invitations).Error; err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error":   "Failed to load invitations",
			"message": err.Error(),
		})
	}

	return c.JSON(invitations)
}

// ResendInvitation resends an invitation email
func (h *TeamsHandler) ResendInvitation(c fiber.Ctx) error {
	userSession := c.Locals("session").(storage.Session)
	invitationID := c.Params("id")

	// Find the invitation
	var invitation models.Invitation
	if err := h.db.
		Where("id = ? AND organization_id = ?", invitationID, userSession.ActiveOrganization).
		First(&invitation).Error; err != nil {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"error":   "Invitation not found",
			"message": "The specified invitation does not exist",
		})
	}

	// Check if the invitation is still pending
	if invitation.Status != models.InvitationPending {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error":   "Invalid invitation status",
			"message": "Only pending invitations can be resent",
		})
	}

	// Check if the invitation has expired and update if needed
	if invitation.ExpiresAt.Before(time.Now()) {
		invitation.ExpiresAt = time.Now().Add(7 * 24 * time.Hour) // 7 days
		if err := h.db.Save(&invitation).Error; err != nil {
			return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
				"error":   "Failed to update invitation",
				"message": err.Error(),
			})
		}
	}

	// TODO: Send invitation email with the token link

	return c.JSON(invitation)
}

// CancelInvitation cancels a pending invitation
func (h *TeamsHandler) CancelInvitation(c fiber.Ctx) error {
	userSession := c.Locals("session").(storage.Session)
	invitationID := c.Params("id")

	// Find the invitation
	var invitation models.Invitation
	if err := h.db.
		Where("id = ? AND organization_id = ?", invitationID, userSession.ActiveOrganization).
		First(&invitation).Error; err != nil {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"error":   "Invitation not found",
			"message": "The specified invitation does not exist",
		})
	}

	// Delete the invitation
	if err := h.db.Delete(&invitation).Error; err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error":   "Failed to delete invitation",
			"message": err.Error(),
		})
	}

	return c.SendStatus(fiber.StatusNoContent)
}

// ValidateInvitation checks if an invitation token is valid and returns invitation details
func (h *TeamsHandler) ValidateInvitation(c fiber.Ctx) error {
	token := c.Params("token")

	// Find the invitation by token
	var invitation models.Invitation
	if err := h.db.
		Preload("Organization").
		Preload("InvitedBy").
		Where("token = ?", token).
		First(&invitation).Error; err != nil {
		return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
			"error":   "Invalid invitation",
			"message": "The invitation token is invalid or has been used",
		})
	}

	// Check if the invitation has expired
	if invitation.ExpiresAt.Before(time.Now()) {
		invitation.Status = models.InvitationExpired
		h.db.Save(&invitation) // Non-critical error, just try to update status

		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error":   "Invitation expired",
			"message": "This invitation has expired",
		})
	}

	// Check if the invitation has already been accepted or rejected
	if invitation.Status != models.InvitationPending {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error":   "Invalid invitation status",
			"message": "This invitation has already been " + string(invitation.Status),
		})
	}

	return c.JSON(invitation)
}

File: packages/internal-api/src/routes/teams/invitations.ts
import { type ErrorResponse, type Opts, createClient } from "@/index";
import type { OtherUser } from "@/routes/auth/basic";
import type { Organization } from "@/routes/teams/organizations";
import { z } from "zod";

export type InvitationStatus = "pending" | "accepted" | "rejected" | "expired";

export interface Invitation {
	id: string;
	created_at: string;
	updated_at: string;
	deleted_at: string | null;
	email: string;
	organization_id: string;
	token: string;
	organization: Organization;
	role: "owner" | "admin" | "member";
	expires_at: string;
	status: InvitationStatus;
	invited_by_id: string;
	invited_by: OtherUser;
}

export const createInvitationSchema = z.object({
	email: z.string().email("Please enter a valid email address"),
	role: z.enum(["admin", "member"], {
		errorMap: () => ({ message: "Invalid role" }),
	}),
});

export type CreateInvitationRequest = z.infer<typeof createInvitationSchema>;

export interface CreateInvitationResponse extends Invitation {
	invite_link?: string; // TODO: Potential future feature
}

export async function createInvitation(
	body: CreateInvitationRequest,
	{ $fetch, ...opts }: Opts,
) {
	const client = $fetch ?? createClient();

	return await client<CreateInvitationResponse, ErrorResponse>(
		"/v1/invitations",
		{
			method: "POST",
			body: JSON.stringify(body),
			credentials: "include",
			...opts,
		},
	);
}

export async function listInvitations({ $fetch, ...opts }: Opts) {
	const client = $fetch ?? createClient();

	return await client<Invitation[], ErrorResponse>("/v1/invitations", {
		credentials: "include",
		...opts,
	});
}

export async function cancelInvitation(
	invitationId: string,
	{ $fetch, ...opts }: Opts,
) {
	const client = $fetch ?? createClient();

	return await client<void, ErrorResponse>(`/v1/invitations/${invitationId}`, {
		method: "DELETE",
		credentials: "include",
		...opts,
	});
}

export async function resendInvitation(
	invitationId: string,
	{ $fetch, ...opts }: Opts,
) {
	const client = $fetch ?? createClient();

	return await client<Invitation, ErrorResponse>(
		`/v1/invitations/${invitationId}/resend`,
		{
			method: "POST",
			credentials: "include",
			...opts,
		},
	);
}

export async function validateInvitation(
	token: string,
	{ $fetch, ...opts }: Opts,
) {
	const client = $fetch ?? createClient();

	return await client<Invitation, ErrorResponse>(
		`/v1/auth/invitations/${token}`,
		{
			credentials: "include",
			...opts,
		},
	);
}

export const acceptInvitationSchema = z.object({
	token: z.string().min(1, "Token is required"),
});

export type AcceptInvitationRequest = z.infer<typeof acceptInvitationSchema>;

export async function acceptInvitation(
	body: AcceptInvitationRequest,
	{ $fetch, ...opts }: Opts,
) {
	const client = $fetch ?? createClient();

	return await client<
		{ message: string; organization: Organization },
		ErrorResponse
	>("/v1/invitations/accept", {
		method: "POST",
		body: JSON.stringify(body),
		credentials: "include",
		...opts,
	});
}

export const acceptInvitationWithRegistrationSchema = z.object({
	token: z.string().min(1, "Token is required"),
	name: z.string().min(1, "Name is required"),
	password: z.string().min(8, "Password must be at least 8 characters"),
});

export type AcceptInvitationWithRegistrationRequest = z.infer<
	typeof acceptInvitationWithRegistrationSchema
>;

export async function acceptInvitationWithRegistration(
	body: AcceptInvitationWithRegistrationRequest,
	{ $fetch, ...opts }: Opts,
) {
	const client = $fetch ?? createClient();

	return await client<{ user: OtherUser; session: any }, ErrorResponse>(
		"/v1/auth/invitations/accept-with-registration",
		{
			method: "POST",
			body: JSON.stringify(body),
			credentials: "include",
			...opts,
		},
	);
}
